Remove `balance` and `devices` parameter from Pipe.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47829 Modify Pipe to return an RRef.
* **#46804 Remove `balance` and `devices` parameter from Pipe.**

As per our design in https://github.com/pytorch/pytorch/issues/44827,
changign the API such that the user places modules on appropriate devices
instead of having a `balance` and `devices` parameter that decides this.

This design allows us to use RemoteModule in the future.

Differential Revision: [D24524219](https://our.internmc.facebook.com/intern/diff/D24524219/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24524219/)!

pritamdamania committed:
Remove `balance` and `devices` parameter from Pipe.

As per our design in https://github.com/pytorch/pytorch/issues/44827,
changign the API such that the user places modules on appropriate devices
instead of having a `balance` and `devices` parameter that decides this.

This design allows us to use RemoteModule in the future.

Differential Revision: [D24524219](https://our.internmc.facebook.com/intern/diff/D24524219/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24524219/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Remove `balance` and `devices` parameter from Pipe."


As per our design in https://github.com/pytorch/pytorch/issues/44827,
changign the API such that the user places modules on appropriate devices
instead of having a `balance` and `devices` parameter that decides this.

This design allows us to use RemoteModule in the future.

Differential Revision: [D24524219](https://our.internmc.facebook.com/intern/diff/D24524219/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24524219/)!

[ghstack-poisoned]

mrshenli commented on test/distributed/_pipeline/sync/test_pipe.py:
```diff
@@ -525,46 +485,55 @@ def test_named_children():
+            ' to place the module on a single device'):
+        Pipe(model)
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="cuda required")
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="Need atleast two GPUs")
```
curious, doesn't existing unittest skip helpers work? I did use those locally with pytest and seems working for me.

mrshenli commented on torch/distributed/_pipeline/sync/pipe.py:
```diff
@@ -89,66 +89,46 @@ class BalanceError(ValueError):
-    Returns:
-        A tuple of (partitions, balance, devices).
+    Arguments:
+        An nn.Module to process.
```
double quote code

mrshenli commented on torch/distributed/_pipeline/sync/pipe.py:
```diff
@@ -89,66 +89,46 @@ class BalanceError(ValueError):
-        item is a partition. All layers in a partition are placed in the
-        same device.
+    Returns:
+        torch.Device for the entire module.
```
ditto

mrshenli commented on torch/distributed/_pipeline/sync/pipe.py:
```diff
@@ -89,66 +89,46 @@ class BalanceError(ValueError):
-            the number of devices is fewer than the number of partitions.
-
+        ValueError:
+            If devices for nn.Module parameters are not all same.
```
ditto

mrshenli commented on torch/distributed/_pipeline/sync/pipe.py:
```diff
@@ -160,28 +140,27 @@ class Pipe(Module):
-    into multiple devices according to the given balance. You may rely on
-    heuristics to find your own optimal configuration.
+    You should place all the modules on the appropriate devices before passing
+    them to this API and wrap them into an nn.Sequential module defining the
```
ditto

mrshenli commented on torch/distributed/_pipeline/sync/pipe.py:
```diff
@@ -160,28 +140,27 @@ class Pipe(Module):
+            Sequential module to be parallelized using pipelining. Each module
+            in the sequence has to have all of its parameters on a single
+            device. Each module in the sequence has to either be an nn.Module
+            or nn.Sequential (to combine multiple sequential modules on a single
```
ditto for this block

mrshenli commented on torch/testing/_internal/distributed/pipeline/utils.py:
```diff
@@ -0,0 +1,21 @@
+from torch import nn
+from typing import List
+
+def convert_to_balance(pipe: nn.Sequential, balance: List[int]):
```
any reason for having a second `convert_to_balance` in test folder?

mrshenli reviewed (approved):
LGTM! Added some minor comments.

pritamdamania committed:
Update on "Remove `balance` and `devices` parameter from Pipe."


As per our design in https://github.com/pytorch/pytorch/issues/44827,
changign the API such that the user places modules on appropriate devices
instead of having a `balance` and `devices` parameter that decides this.

This design allows us to use RemoteModule in the future.

Differential Revision: [D24524219](https://our.internmc.facebook.com/intern/diff/D24524219/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24524219/)!

[ghstack-poisoned]

ngimel commented:
Reverting, broke flake8
